### PR TITLE
Added `_register_any_hook`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115629
* __->__ #115628

This PR adds a function `_register_any_hook` that registers a backward tensor hook that runs when _any_ of the passed-in tensors' gradients are computed and becomes a no-op thereafter. This is used to implement the pre-backward hook in FSDP that all-gathers parameters needed for backward computation.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -439,27 +440,60 @@ def inner_hook(grad: torch.Tensor):
...
+            fn(*args, **kwargs)
+        return
+
+    handles = tuple(
+        tensor.register_hook(wrapped_fn) for tensor in tensors if tensor.requires_grad
+    )
+    return _MultiHandle(handles)
+
+
+class _MultiHandle(RemovableHandle):
```
Comment (awgu) [inline torch/autograd/graph.py:482]: Previously, the `Handle` class was defined only in `register_multi_grad_hook()`. However, `Handle` was not public, so I kept this `_MultiHandle` private for now.

File: torch/autograd/graph.py [hunk 2]
```diff
@@ -439,27 +440,60 @@ def inner_hook(grad: torch.Tensor):
...
+
+    This function returns a handle with a method ``handle.remove()`` that removes the hook.
+    """
+    ran_hook = False
+
+    @functools.wraps(fn)
+    def wrapped_fn(*args, **kwargs):
+        nonlocal ran_hook
+        if not ran_hook:
+            ran_hook = True
```
Comment (awgu) [inline torch/autograd/graph.py:472]: IIUC, this is not thread-safe, as we would need to ensure atomicity of the read and write to `ran_hook`. For the FSDP use-case, we do not need thread safety since we do not run backward with multiple threads. Should we make this thread-safe, or should we just document that it is not thread safe?

File: torch/autograd/graph.py [hunk 3]
```diff
@@ -439,27 +440,60 @@ def inner_hook(grad: torch.Tensor):
...
+    return _MultiHandle(tuple(handles))
 
-        def remove(self):
-            for handle in self.handles:
-                handle.remove()
 
-        def __getstate__(self):
-            return self.handles
+def _register_any_hook(
+    tensors: Sequence[torch.Tensor],
```
Comment (awgu) [inline torch/autograd/graph.py:451]: We have this hook only take in `Sequence[torch.Tensor]` and leave extracting the `torch.Tensor`s (that require gradient) from the forward outputs to FSDP.

**Review by awgu — COMMENTED** [review]
@albanD Let me know if this should go somewhere other than `torch/autograd/graph.py`, if we should consider making this public, and/or if you prefer a different name.

This is mainly for (arguably) simplifying FSDP code to avoid maintaining the `bool` to ensure the hook only runs once.

File: torch/autograd/graph.py (hunk 2 above)
Comment (soulitzer) [reply torch/autograd/graph.py:472]: Maybe do something like the following?

```python
# outside hook
lock = threading.lock()
ran_hook = default_dict(bool)

# inside hook
id = torch._C._current_graph_task_id()
assert id != -1, "Expect hook to be executed during backward"

lock.acquire()
prev, ran_hook[id] = ran_hook[id], True
lock.release()
if prev:
    return
fn(*args, **kwargs)
```

File: torch/autograd/graph.py (hunk 3 above)
Comment (soulitzer) [reply torch/autograd/graph.py:451]: Hmm wdym, we also check if tensors that require grad below?

Comment (awgu) [reply torch/autograd/graph.py:451]: Sorry for the confusion. I mainly meant that we do not allow arbitrary Python object as input to this function but rather just `Sequence[torch.Tensor]`:
https://github.com/pytorch/pytorch/blob/c19184cb61d12aea143586e776c5037bcb6b1022/torch/distributed/fsdp/_runtime_utils.py#L1377-L1383

Comment (soulitzer) [reply torch/autograd/graph.py:451]: Ah gotcha, thanks for clarifying!

**Review by soulitzer — COMMENTED** [review]
> @albanD Let me know if this should go somewhere other than torch/autograd/graph.py, if we should consider making this public, and/or if you prefer a different name.

If we are putting in the `autograd.graph` namespace, it would be best if this were to be public and live under the `multi_tensor_hook` as a mode (otherwise if we also have something like any_hook, it could be a little confusing what multi tensor hook does.
We could accept `mode` as a kwarg-only str where (`"all"` , `"first"`) are the allowed options, and `"all"` is the default for bc.

**Comment by soulitzer** [comment]
> @soulitzer Just to check, do you mean under register_multi_grad_hook (instead of multi_tensor_hook)?

Ah yes that is what I mean :)

**Comment by awgu** [comment]
@soulitzer I think your suggestion sounds good! Let me wait for @albanD to leave a comment, and I can make the API change plus add the logic for thread-safety.

**Comment by albanD** [comment]
Yes making this properly public sounds good!

**Comment by awgu** [comment]
This is a bit tricky. For existing FSDP, we cannot use this hook as is. We need the `ran_hook: bool` that ensures that the hook only runs once to apply across multiple registrations -- i.e. we need the `bool` to be per hook function, not per call to `_register_any_hook()`. This is because existing FSDP has only a single (pre-backward, post-backward) interval in the case of a module participating multiple times in forward, as opposed to one (pre-backward, post-backward) interval for each participation.

I need to double check that this any hook works for per-parameter FSDP. Conceptually, it should due to following the "one (pre-backward, post-backward) interval for each participation" approach instead of the single interval.

**Comment by awgu** [comment]
Closing in favor of the new PR.
